pkg/route/controller/ingress/ingress.go

```diff
@@ -583,14 +583,16 @@ func targetPortForService(namespace string, path *extensionsv1beta1.HTTPIngressP
 		expect := path.Backend.ServicePort.StrVal
 		for _, port := range service.Spec.Ports {
 			if port.Name == expect {
-				return &port.TargetPort
+				targetPort := intstr.FromString(port.Name)
+				return &targetPort
 			}
 		}
 	} else {
 		for _, port := range service.Spec.Ports {
 			expect := path.Backend.ServicePort.IntVal
 			if port.Port == expect {
-				return &port.TargetPort
+				targetPort := intstr.FromInt(int(port.Port))
+				return &targetPort
```

Comment:
This change is breaking the test in Origin.  The router uses this target port to look up the port in the _endpoints_ resource, not the _service_ resource, and although a port in a service resource and the corresponding port in the endpoints resource have the same _name_, the port _number_ in the endpoints resource comes from the _target port_ of the service.  For example, the following service resource:
```yaml
kind: Service
# ...
spec:
  ports:
  - name: http-default
    port: 8888
    targetPort: 8080
```
will have the following endpoints resource:
```yaml
kind: Endpoints
# ...
subsets:
- # ...
  ports:
  - name: http-default
    port: 8080
```
The following route resource will match the endpoints resource:
```yaml
kind: Route
# ...
spec:
  # ...
  port:
    targetPort: 8080
```
and so will the following:
```yaml
kind: Route
# ...
spec:
  # ...
  port:
    targetPort: http-default
```
but the following will not:
```yaml
kind: Route
# ...
spec:
  # ...
  port:
    targetPort: 8888
```

Reply:
OK - now I understand. The original bug #23937 only happens when the targetPort of the Service is a named (string) port of the Pod. When the targetPort is an int the endpointPort is already correct in the current implementation.